experimental/c-cpp/build-generator.py

```diff
@@ -0,0 +1,1486 @@
+  build_script += "rm -rf /%s\n" % (test_dir)
+  build_script += "cp -rf %s %s\n" % (abspath_of_target, test_dir)
+  build_script += "cd %s\n" % (test_dir)
+  for cmd in build_container.list_of_commands:
+    build_script += cmd + "\n"
+
+  return build_script
+
+
+def get_all_functions_in_project(introspection_files_found):
+  all_functions_in_project = []
+  for fi_yaml_file in introspection_files_found:
+    with open(fi_yaml_file, "r") as file:
+      yaml_content = yaml.safe_load(file)
+    for elem in yaml_content['All functions']['Elements']:
+      all_functions_in_project.append(elem)
+
+  return all_functions_in_project
+
+
+##################################################
+#### Heuristics for auto generating harnesses ####
+##################################################
+
+GLOBAL_FUZZER_SOURCE_CACHE = {}
+
+
+def get_source_from_cache(heuristic_name, target_func):
+  funcs_in_cache = GLOBAL_FUZZER_SOURCE_CACHE.get(heuristic_name, [])
+  if len(funcs_in_cache) == 0:
+    return None
+  for func, target_source in funcs_in_cache:
+    if func['Func name'] == target_func['Func name']:
+      return target_source
+  return None
+
+
+def add_to_source_cache(heuristic_name, target_func, fuzzer_source):
+  funcs_in_cache = GLOBAL_FUZZER_SOURCE_CACHE.get(heuristic_name, [])
+  funcs_in_cache.append((target_func, fuzzer_source))
+  GLOBAL_FUZZER_SOURCE_CACHE[heuristic_name] = funcs_in_cache
+
+
+class FuzzHeuristicGeneratorBase:
+  """Base class for fuzzer heuristics generator."""
+
+  def __init__(self, test_dir):
+    self.test_dir = test_dir
+    self.all_header_files = []
+    self.all_functions_in_project = []
+    self.introspector_report = {}
+
+  @abstractmethod
+  def get_fuzzer_intrinsics(self, func) -> Dict[str, Any]:
+    """generates fuzzer source code, build and include directives."""
+
+  @abstractmethod
+  def get_fuzzing_targets(self) -> List[Any]:
+    """Gets a list of possible function targets."""
+
+  def get_header_intrinsics(self):
+    """All header files and include directories."""
+    headers_to_include = set()
+    header_paths_to_include = set()
+    for header_file in self.all_header_files:
+      #print("- %s"%(header_file))
+      if "/test/" in header_file:
+        continue
+      if "googletest" in header_file:
+        continue
+      headers_to_include.add(os.path.basename(header_file))
+      header_paths_to_include.add("/".join(header_file.split("/")[1:-1]))
+
+    # Generate -I strings to be used in the build command.
+    build_command_includes = ""
+    for header_path_to_include in header_paths_to_include:
+      build_command_includes += "-I" + os.path.join(
+          self.test_dir, header_path_to_include) + " "
+
+    return headers_to_include, header_paths_to_include, build_command_includes
+
+  def run_prompt_and_get_fuzzer_source(self, prompt):
+    """Communicate to OpenAI prompt and extract harness source code."""
+    completion = client.chat.completions.create(model="gpt-3.5-turbo",
```

Comment:
Can you add a TODO to consolidate this with our prompt calling code?

i.e. https://github.com/google/oss-fuzz-gen/blob/main/llm_toolkit/prompts.py and https://github.com/google/oss-fuzz-gen/blob/620dec55cab5c78badd66c1b23f3edeafd0cda58/llm_toolkit/models.py#L122

Reply:
Yes, done.